### PR TITLE
WIP Feature/add non root user in dev container

### PR DESCRIPTION
**Pull request by Jason5480**
Make docker dev container safer to use by logging to a non-root user as described here https://aka.ms/vscode-remote/containers/non-root .
Clean up some defaulted arguments.
Install cmake_format in the container

### CODE REVIEW

**Comment by Jason5480** [comment]
> Unfortunately, this causes some low level conan package installation to fail since they are dependent on system packages and somehow the installation does not work for non-root users... @ddalcino did you face that also in the past during the very first implementation? `CONAN_SYSREQUIRES_MODE`=enabled doesn't seem to work for regular users.

The errors are not caused by the user changes of this branch. I went in the `main` branch and a new dependency was needed (pkg-config) when I added that, I got the following error

image

I will create a separate issue for that

**Comment by lefticus** [comment]
@Jason5480 I'm currently trying to reorganize things into a light-weight "boilerplate" project and this "starter_project"

Could I trouble you to move this PR over to https://github.com/cpp-best-practices/cpp_boilerplate_project, since that is the new baseline project, and we'll then merge the changes back into the starter_project?

**Comment by ddalcino** [comment]
> Unfortunately, this causes some low level conan package installation to fail since they are dependent on system packages and somehow the installation does not work for non-root users... @ddalcino did you face that also in the past during the very first implementation? `CONAN_SYSREQUIRES_MODE`=enabled doesn't seem to work for regular users.

`CONAN_SYSREQUIRES_MODE`=enabled means that the Conan process does not need superuser privileges, because it already has them. Many Conan packages have dependencies that must be installed via the system package manager, and typically, sudo is required for this. Without `CONAN_SYSREQUIRES_MODE`=enabled, you will need to `apt-get install sudo`. I don't know how to make this work without superuser privileges, unless you stick to packages that don't need it. That means no GUI libraries like SDL.

You could try going through each packages' dependencies manually, and installing them yourself; I've done this in the past, and gotten SDL to build entirely without superuser privileges. It was ridiculous, and I'll never do it again. I do not recommend this approach.

**Comment by lefticus** [comment]
> > Unfortunately, this causes some low level conan package installation to fail since they are dependent on system packages and somehow the installation does not work for non-root users... @ddalcino did you face that also in the past during the very first implementation? `CONAN_SYSREQUIRES_MODE`=enabled doesn't seem to work for regular users.
> 
> `CONAN_SYSREQUIRES_MODE`=enabled means that the Conan process does not need superuser privileges, because it already has them. Many Conan packages have dependencies that must be installed via the system package manager, and typically, sudo is required for this. Without `CONAN_SYSREQUIRES_MODE`=enabled, you will need to `apt-get install sudo`. I don't know how to make this work without superuser privileges, unless you stick to packages that don't need it. That means no GUI libraries like SDL.
> 
> You could try going through each package's dependencies manually, and installing them yourself; I've done this in the past, and gotten SDL to build entirely without superuser privileges. It was ridiculous, and I'll never do it again. I do not recommend this approach.

Thank's @ddalcino for your insight - I really know nothing about container usage in this way

**Comment by Jason5480** [comment]
> @Jason5480 I'm currently trying to reorganize things into a light-weight "boilerplate" project and this "starter_project"
> 
> Could I trouble you to move this PR over to https://github.com/cpp-best-practices/cpp_boilerplate_project, since that is the new baseline project, and we'll then merge the changes back into the starter_project?

I created that one https://github.com/cpp-best-practices/cpp_boilerplate_project/pull/8 where I added some missing format tools into the container. @ddalcino I will reconsider the non-root user solution and I will try to find a solution to work with conan (if any). Using root user in containers is not accepted in many companies for security reasons...

**Comment by Jason5480** [comment]
> @Jason5480 I'm currently trying to reorganize things into a light-weight "boilerplate" project and this "starter_project"
> 
> Could I trouble you to move this PR over to https://github.com/cpp-best-practices/cpp_boilerplate_project, since that is the new baseline project, and we'll then merge the changes back into the starter_project?

I created that one https://github.com/cpp-best-practices/cpp_boilerplate_project/pull/8 where I added some missing format tools to the container. @ddalcino I will reconsider the non-root user solution and I will try to find a solution to work with conan (if any). Using root user in containers in not accepted in many companies for security reasons

**Comment by Jason5480** [comment]
> > Unfortunately, this causes some low level conan package installation to fail since they are dependent on system packages and somehow the installation does not work for non-root users... @ddalcino did you face that also in the past during the very first implementation? `CONAN_SYSREQUIRES_MODE`=enabled doesn't seem to work for regular users.
> 
> `CONAN_SYSREQUIRES_MODE`=enabled means that the Conan process does not need superuser privileges, because it already has them. Many Conan packages have dependencies that must be installed via the system package manager, and typically, sudo is required for this. Without `CONAN_SYSREQUIRES_MODE`=enabled, you will need to `apt-get install sudo`. I don't know how to make this work without superuser privileges, unless you stick to packages that don't need it. That means no GUI libraries like SDL.
> 
> You could try going through each package's dependencies manually, and installing them yourself; I've done this in the past, and gotten SDL to build entirely without superuser privileges. It was ridiculous, and I'll never do it again. I do not recommend this approach.

Do you know any command to check if the current user is a superuser and set those variables accordingly?

**Comment by ddalcino** [comment]
> Do you know any command to check if the current user is a superuser and set those variables accordingly?

You can try checking $EUID == 0, or google it. Personally, I think I would prefer an option that I can set (possibly an ARG) that controls whether or not this runs in superuser mode or not, and set the default to non-superuser. Projects that use SDL etc. would have to either use the option, change the default, or install the dependency outside of Conan.

Thanks for looking into this PR, this is a serious security issue. I hated writing the original in this way, but I didn’t see an alternative.

**Comment by ddalcino** [comment]
> `CONAN_SYSREQUIRES_MODE`=enabled means that the Conan process does not need superuser privileges, because it already has them. Many Conan packages have dependencies that must be installed via the system package manager, and typically, sudo is required for this. Without `CONAN_SYSREQUIRES_MODE`=enabled, you will need to `apt-get install sudo`. I don't know how to make this work without superuser privileges, unless you stick to packages that don't need it. That means no GUI libraries like SDL.

Whoops, looks like I got this backwards. I was thinking of CONAN_SYSREQUIRES_SUDO; see https://docs.conan.io/en/latest/reference/env_vars.html?highlight=conan_sysrequires_mode#conan-sysrequires-sudo.
CONAN_SYSREQUIRES_MODE is documented here: https://docs.conan.io/en/latest/reference/env_vars.html?highlight=conan_sysrequires_mode#conan-sysrequires-mode

If I had read the comments I left in the Dockerfile, I would not have made this mistake.

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -21,7 +32,7 @@ RUN python3 -m pip install --upgrade pip setuptools && \
 # By default, anything you run in Docker is done as superuser.
 # Conan runs some install commands as superuser, and will prepend `sudo` to
 # these commands, unless `CONAN_SYSREQUIRES_SUDO=0` is in your env variables.
-ENV CONAN_SYSREQUIRES_SUDO 0
+ENV CONAN_SYSREQUIRES_SUDO 1
 # Some packages request that Conan use the system package manager to install
 # a few dependencies. This flag allows Conan to proceed with these installations;
 # leaving this flag undefined can cause some installation failures.
```
Comment (ddalcino) [inline .devcontainer/Dockerfile:38]:
```suggestion
# See docs at https://docs.conan.io/en/latest/reference/env_vars.html?highlight=conan_sysrequires_mode#conan-sysrequires-sudo
ENV CONAN_SYSREQUIRES_SUDO 1
# Some packages request that Conan use the system package manager to install
# a few dependencies. This flag allows Conan to proceed with these installations;
# leaving this flag undefined can cause some installation failures.
# See docs at https://docs.conan.io/en/latest/reference/env_vars.html?highlight=conan_sysrequires_mode#conan-sysrequires-mode
```

Comment (Jason5480) [reply .devcontainer/Dockerfile:38]: I will add this here https://github.com/Jason5480/cpp_boilerplate_project/tree/feature/add_non_root_user_in_docker_container I will close that MR for now as I will not procced any further. I will make a new MR when I find how I can make it work.

**Pull request closed without merging**